fossology/fossology
# fix(license-csv): Update license if exists

## Description

With the current implementation, if a license already exists in the DB where CSV is being imported, it is skipped altogether. And if the license does not exists, but it's MD5 matches with other license, the other license is updated with new license.

This PR fixes this behavior as following:
1. If the license shortname already exists
    1. If the MD5 of new license does not collide with any other license, update the license else print an error.
1. If the license shortname does not exists
    1. If the MD5 of the new license collides with any other license, print an error.
    1. Otherwise insert it as a new license.

This PR also provides way to differentiate candidate licenses with main licenses by creating additional column called `group` in the exported CSV and checking this column while importing the license to determine if the license is candidate or not.
1. If the row contains `group` value, it is candidate.
1. The MD5 collisions will not be checked.
1. If the group does not exists in current DB, skip the license and print an error.

Also, while exporting the licenses, do not export candidate licenses which do not have merge request open (`marydone`). And while importing licenses, open merge request for all candidate licenses.

## How to test

1. Export your license list with modified licenses, new licenses, candidate licenses with and without merge request.
1. Examine if the exported CSV have groups against candidate licenses and nothing against main licenses.
1. Import the list on different server (with some licenses).
    1. Check if existing licenses are updated.
    1. Check if new licenses are created.
    1. Check if candidate licenses are behaving as expected.
1. Import an old export (without `group` column), the import should not break.
